### PR TITLE
BLD: update requirements to use cython>3.0

### PR DESCRIPTION
**Pull request by mattip**
Cython3.0 has been released. Let's update to use it everywhere. For builds I pinned an upper limit `<3.1`, but for tests we can try the latest release.

### CODE REVIEW

**Comment by mattip** [comment]
Bad news. Cython3 (mis)uses the `NO_IMPORT_ARRAY` macro to inject a call to `import_array`. The CI job we have for Cython3 somehow is not using the `-Werror=undef` flag which comes from `sysconfig.get_config_var("CFLAGS")`, maybe because it was attached to the debug build which was built with different flags. I opened cython/cython#5557 to fix this upstream.

**Comment by rgommers** [comment]
Thanks for moving this forward Matti! I'm +1 to moving to Cython >=3 in `main`, the 2.0 timeline is fine for that.

In case anyone was thinking about 1.25.x/1.26.x: for those releases I think we should avoid this, because there are distros without build isolation, so a build requirement can be forced upon a whole distro at once, and many packages need time to update to Cython 3.x compatibility.

**Comment by rgommers** [comment]
> For builds I pinned an upper limit `<3.1`, but for tests we can try the latest release.

We should never do this on `main` except to work around known breakage.

File: pyproject.toml.setuppy
Comment (charris) [inline pyproject.toml.setuppy:8]: Do you intend pinning the upper value?

File: tools/travis-test.sh
Comment (charris) [inline tools/travis-test.sh:64]: Still need setup here?

Comment (rgommers) [reply tools/travis-test.sh:64]: Yes, because the TravisCI jobs are for testing SIMD features, so we should only port them over once the Meson SIMD support lands (that is much closer now, but not there yet).

File: pyproject.toml.setuppy
Comment (rgommers) [reply pyproject.toml.setuppy:8]: This seems okay to me, since this is a fallback file for switching back to `setup.py` based builds via `pip`. For releases we do tend to want upper bounds there.

File: environment.yml
Comment (rgommers) [inline environment.yml:18]: not sure where this one came from, we already require >=0.13.1 in `pyproject.toml` - better to leave unpinned, there no use for a lower bound here

Let's leave `pip` I'd say, it's needed a lot for development, not only for installing `spin` but for, e.g., editable installs.

File: environment.yml
Comment (rgommers) [inline environment.yml:16]: this one is good indeed, it's a dependency of meson-python and that won't change.

File: environment.yml
Comment (rgommers) [inline environment.yml:19]: no need to pin this one, I'd expect new releases to come with functionality and fixes we need

**Review by rgommers — APPROVED** [review]
LGTM modulo the few tweaks to `environment.yml` that I suggested. Thanks Matti

File: environment.yml
Comment (mattip) [reply environment.yml:16]: Do you mean "good, it is a dependency and so this line is not needed" or "good, please revert the removal"?

Comment (rgommers) [reply environment.yml:16]: sorry, "good it is indeed not needed". I'll resolve this one.

**Comment by seberg** [comment]
Awesome, let's give it a shot, thanks!  Hopefully that will also help cython 3 a bit.

Should we have a milestoned issue to remove the undef warning filter?  I guess it's pretty harmless overall.

**Comment by mattip** [comment]
> Should we have a milestoned issue to remove the undef warning filter?

Maybe? It is part of the CFLAGS used by the non-meson builds, so it will probably get reviewed when we refactor those builds to use meson.

**Pull request merged**
